---
title: "Class 18"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

>P53_wt Cellular tumor antigen p53 - Healthy Tissue
MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP
DEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAK
SVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHE
RCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNS
SCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELP
PGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPG
GSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD

>P53_mutant Cellular tumor antigen p53 - Tumor Tissue
MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMLDLMLSPDDIEQWFTEDPGP
DEAPWMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAK
SVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHE
RCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFVHSVVVPYEPPEVGSDCTTIHYNYMCNS
SCMGGMNRRPILTIITLEV


What are the tumor specific amino-acids?
what peptides include these amino acids

Loading the required packages
```{r}
library(bio3d)
```

After aligning my two sequences with seaview, I will load in my aligned Fasta sequences
```{r}
fastas <- read.fasta("lecture18_sequences.fa")
fastas
```

However, I could have more easily utilized the function seqaln() from the bio3D package...
```{r}
#aln <- seqaln(fastas)

#aln
```



we want to pull the 8 aa's to either side of the mismatches... we'd need both the sequence clips
identify it they match

```{r}
fastas$id  #this produces the protein sequence names

fastas$call #this reminds me how I called on this file

ali <- fastas$ali #this is the information in workable form ...
```

```{r}
gaps <- gap.inspect(fastas)
gap.inds <- gaps$t.inds
```


```{r}
ide <- conserv(fastas, method="identity")
mismatch.inds <- which(ide < 1)
```

```{r}
tumor.sites <- mismatch.inds[!mismatch.inds %in% gap.inds]

fastas$ali[,tumor.sites]
```

```{r}
ids <- paste(fastas$ali[1,tumor.sites], 
     tumor.sites, 
     fastas$ali[2,tumor.sites], sep="")

ids
```

```{r}
start.ind <- tumor.sites - 8
end.ind <- tumor.sites + 8

tumor <- NULL
for(i in 1:length(start.ind)) {
    tumor <- seqbind(tumor, 
                     fastas$ali[2,start.ind[i]: end.ind[i]])
}

tumor

write.fasta(ids=ids, ali=tumor, file="subsequences.fa")

```
















